Controlsys package

benepla opened the pull request:
Adds a simple package for symbolic handling of linear control systems. The package was written during an internship at the RWTH Aachen, while porting Mathematica features into sympy.
Models can be transformed in state space or transfer function realization, can be interconnected and evaluated in a simple way.
Evaluating can be done symbolically but takes many integrations (n*m with n outputs and m inputs) and will probably be slow in most cases (but can be very useful).
Numerical evaluation uses scypy, because mpmath numerical integration is very slow. Is this allowed?
##### TODO (updated):
###### Testing:
- [x] Added a unittest folder and tests for all features
- [x] Add doctests/examples
- [x] Add test for numerical evaluation or dismiss this feature completely.
###### Refining usage:
- [x] Change API from StateSpaceModel([A, B, C, D]) to StateSpaceModel(A, B, C, D)
- [x] Added constructors for calls with less then four matrices (so StateSpaceModel(), StateSpaceModel(A) etc.) This can be very convenient when only testing for controllability.
- [x] Added comparison between systems.
  The package was only used by me for now, so I don't know if it is convenient to use as it is now.
## Usage
### Creating models
##### From Matrices:

You can create a 'StateSpaceModel' object using four matrices (state space representation) or a or 'TransferFunctionModel' using a transfer matrix:

``` python
from sympy import *     # needed for matrix class

var('a:d')
A, B, C, D = Matrix([a]), Matrix([b]), Matrix([c]), Matrix([d])

var('s)
T = Matrix([1 / s, s / (1 + s**2)], s)

ssm = StateSpaceModel(A, B, C, D)
tfm = TransferFunctionModel(T)
```

One can also just use only one, two or three matrices. The missing ones will be filled with a minimum of zeros.

``` python
var('a:2, b:2')
ssm = StateSpaceModel(Matrix([a0, a1]), Matrix([b0, b1]))
```
##### From Other Systems

You can also create the models from one another:

``` python
ssm2 = StateSpaceModel(tfm)
tfm2 = TransferFunctionModel(ssm, s)
```

Or connect two system to create a new one:

``` python
var('a:3, b:3')
ssm1 = StateSpaceModel(
        Matrix([[0, 1],
               [-a0, -a1]]),
        Matrix([0, 1]),
        Matrix([[b0, b1]]))
ssm2 = StateSpaceModel(
        Matrix([[0, 1, 0],
                [0, 0, 1],
                [-a0, -a1, -a2]]),
        Matrix([0, 0, 1]),
        Matrix([[b0, b1, b2]]))
ssm_series = ssm1.cascade(ssm2)
ssm_parallel = ssm2.parallel(ssm2)
```

``` python
var('a, b, s')
tfm1 = TransferFunctionModel(Matrix([1 / (a * s)]))
tfm2 = TransferFunctionModel(Matrix([(b + s) / (a + s)]))
tfm_series = tfm1.cascade(tfm2)
tfm_parallel = tfm1.cascade(tfm2)
```

Note that the systems must have a matching nubmer of inputs and outputs for this to work!
### Evaluating models

Models are always evaluated using the `evaluate` method. It is done numercialy or symbolicaly, depending on the parameters of the call:

``` python
# symbolic evaluation:
var('t')
u = Matrix([Heaviside(t)])
x0 = zeros(1, 1)
t0 = 0
y = ssm.evaluate(u, x0, t, t0)

# numerical evaluation
import numpy as np    # use of numpy array is optional
y_list = ssm.evaluate(u, x0, (t, np.arange(0, 10, 0.1)), t0)
```

Note that the only difference between the calls is that in the second one, t is substituted by `(t, np.arange(0, 10, 0.1))`; a tuple of the symbol t and a list of times to evaluate for. 
The first call will return a matrix valued expression in terms of t.
The second call will return a list of matrix valued outputs, matching the time input.

For details on syntax and additional methods, see the docstings of each class and its methods.

benepla committed:
Added a new controlsys package.

The package implements basic linear, time invariant control systems.
it provides two classes for state space and transfer function model
representation of these systems.

Examples:

create a state space model (SSM) from four matrices

In [1]: var('a:d')
		A, B, C, D =  Matrix([a]), Matrix([b]), Matrix([c]), Matrix([d])
		ssm = StateSpaceModel([A, B, C, D])
		ssm

Out[1]: StateSpaceModel([
		[Matrix([[a]]), Matrix([[b]])],
		[Matrix([[c]]), Matrix([[d]])]])

create a transfer function model (TFM) from one matrix

In [2]: var('s')
		tfm = TransferFunctionModel(Matrix([1/s]))
		tfm

Out[2]: TransferFunctionModel([[1/s]])

create a SSM from a TFM and vice-versa

In [3]: ssm2 = StateSpaceModel(tfm)
		tfm2 = TransferFunctionModel(ssm)
		ssm, tfm

Out[2]: StateSpaceModel([
		[Matrix([[0]]), Matrix([[1]])],
		[Matrix([[1]]), Matrix([[0]])]])
		TransferFunctionModel([[(-b*c + d*(a - s))/(a - s)]])

The package basically implements a former standalone package using
sympy which was written by me very recently. For further examples and
usage see the docstrings of each functions. The old project can be found
here but is not up to date:

https://github.com/m3zz0m1x/LTI-Systems-for-Sympy

benepla committed:
updated the string representation of the models.

They now match the representation stated in thr last commit messeges.
Before the commit, the representation was just the same as of a
blockmatrix containing the state space matrices:

In [1]: str(ssm)
Out[1]: Matrix( .. )

not it would be

Out[1]: StateSpaceModel( .. )

The same goes for TransferFunctionModel class.

benepla committed:
Various small corrections.

Corrected some typos and broken variable references.

moorepants commented:
@m3zz0m1x Thanks for this submission! This is a great idea and very nice work. We'd love to get it merged into SymPy. I'm happy to help review this, but I'm extremely busy at the moment with the start of our school's quarter, so bare with me. There will be some other's that are interested in this too, so they will likely help.

moorepants commented:
Comment #1: You will need to add unit tests that test the functionality of your new classes. Currently there are none. Please checkout the `tests` directories in other SymPy sub-packages for examples.

moorepants commented:
> `A, B, C, D = Matrix([a]), Matrix([b]), Matrix([c]), Matrix([d])`

This notation confuses me. Here are you using the Matrix objects to represent arbitrarily sized matrices? If so, you will want to checkout MatrixSymbol which is our object for representing arbitrary matrices.

moorepants commented:
> `T = Matrix([num /denom], s)`

Why do you choose to use a Matrix object for a transfer function? Would two polynomial objects be preferable?

moorepants commented on sympy/controlsys/lti.py:
```diff
@@ -0,0 +1,757 @@
+    """state space model (ssm) of a linear, time invariant control system
+
+    Represents the standard state-space model with state matrix A, input matrix B, output matrix C, and
+    transmission matrix D. This makes the linear controll system:
```
controll > control

moorepants commented on sympy/controlsys/lti.py:
```diff
@@ -0,0 +1,757 @@
+    See Also
+    ========
+
+    TranferFunctionModel: transfer function model of a lti system
```
Tranfer > Transfer

certik replied:
@m3zz0m1x you forgot to address this comment.

moorepants commented on sympy/controlsys/lti.py:
```diff
@@ -0,0 +1,757 @@
+        else:
+            # store the argument as representation of the system
+            try:
+                self.represent = arg[:4]
```
Would a preferable API be:

`StateSpaceModel(A, B, C, D)` instead of `StateSpaceModel([A, B, C, D])`?

moorepants commented on sympy/controlsys/lti.py:
```diff
@@ -0,0 +1,757 @@
+                    raise ShapeError("Shapes of A,B,C,D must fit")
+
+                # create a block matrix [[A,B], [C,D]] for visual representation
+                self.BlockRepresent = BlockMatrix([[self.represent[0], self.represent[1]],
```
BlockRepresent > blockrepresent, PEP8

You may want this private too: `_blockrepresent`.

moorepants commented on sympy/controlsys/lti.py:
```diff
@@ -0,0 +1,757 @@
+        D = G.limit(s, oo)
+
+        # define G_sp as the (stricly proper) difference of G and D
+        G_sp = simplify(G - D)
```
These last two commands can potentially take a long time for complicated expressions, especially simplify. Simplify is typically an option or user commanded function due to it's potential for being slow. I would leave it out here.

moorepants commented on sympy/controlsys/lti.py:
```diff
@@ -0,0 +1,757 @@
+            C = C.row_join(G_sp_coeff[i])
+
+        # return the state space representation
+        return [simplify(A), simplify(B), simplify(C), simplify(D)]
```
All of the internal simplify functions in this command should be removed. The user can call simplify on the result if they choose.

moorepants commented on sympy/controlsys/lti.py:
```diff
@@ -0,0 +1,757 @@
+            Integral object instead. Only valid for symbolic solutions, ignored otherwise
+        dps : integer
+            the decimal precision of numericial integration
+
```
It would be useful (and typical) to include and "Examples" section in each of these docstrings to show how the class is used. Check out other SymPy classes.

moorepants commented on sympy/controlsys/lti.py:
```diff
@@ -0,0 +1,757 @@
+
+        try:
+
+            # if t symobl, then calculate the solution symbolicaly
```
symobl ? symbol

moorepants commented on sympy/controlsys/lti.py:
```diff
@@ -0,0 +1,757 @@
+        return 'StateSpaceModel' + str(self.BlockRepresent)[6:]
+
+
+class TransferFunctionModel(object):
```
Since a transfer function is a ratio of two polynomials in s you will likely benefit from subclassing from a core sympy type here.

moorepants commented:
>  Numerical evaluation uses scypy, because mpmath numerical integration is very slow. Is this allowed?

SymPy does not have a required dependency on SciPy but we have tools like `lambdify` that generate functions that can be evaluated numerically. The typical way to handle things like this in SymPy is to design you objects such that they all function symbolically and then to use the code generation features to generate fast (usually double precision) code from the symbolics. For example if you want to integrate a state space model you would have your state space model output a symbolic definite integral. Then if you want to evaluate the integral with SciPy you'd pass this integral to lambdify which would generate a function that uses SciPy's linear system integration routine in the signal package.

m3zz0m1x commented:
>> A, B, C, D = Matrix([a]), Matrix([b]), Matrix([c]), Matrix([d]) 

>This notation confuses me. Here are you using the Matrix objects to represent arbitrarily sized matrices? If so, you will want to checkout MatrixSymbol which is our object for representing arbitrary matrices.

I'm sorry this was a confusing example I guess. This was just meant as an example of a general SISO (single input, single output) system. The package does not support MatrixSymbol yet, so you always have to pass four matrices with matching shapes at creation.


 >>T = Matrix([num /denom], s) 

>Why do you choose to use a Matrix object for a transfer function? Would two polynomial objects be preferable?

The fact that a transfer function is a complex function is only true for **SISO** systems, not for **MIMO** (multiple input, multiple output) systems. If you have n inputs and m outputs, you write the inputs in a n dimensional column vector and multiply a (m x n) matrix to obtain a vector with the m outputs. So in general the general MIMO case, we have a **(m x n)-dimensional transfer matrix**. 
A matrix may not be the most efficient data structure (preferable would probably be some nested list of tuples of polys), but a matrix is the easiest to start with I guess.

benepla committed:
fixed typos and removed internal simplification

I was not able to remove the simplify call in
'StateSpaceModel._find_realization' without breaking functionality:

	G_sp_coeff = _matrix_coeff(simplify(G_sp * lcd), s)

Without the simplify, the degree call on the entries will raise an error
in _matrix_degree(). Calling cancel(together(..),extensions=True)
instead of simplify(..) did not work in all cases either.

benepla committed:
Changed API StateSpaceModel([A,B,C,D]) to StateSpaceModel(A,B,C,D).

benepla committed:
updated docstring of StateSpaceModel to reflect new API

m3zz0m1x commented:
>>Numerical evaluation uses scypy, because mpmath numerical integration is very slow. Is this allowed?

>SymPy does not have a required dependency on SciPy but we have tools like  lambdify  that generate functions that can be evaluated numerically. The typical way to handle things like this in SymPy is to design your objects such that they all function symbolically and then to use the code generation features to generate fast (usually double precision) code from the symbolics. For example if you want to integrate a state space model you would have your state space model output a symbolic definite integral. Then if you want to evaluate the integral with SciPy you'd pass this integral to lambdify which would generate a function that uses SciPy's linear system integration routine in the signal package.

okay, what can I do then to prevent sympy from trying to evaluate the matrix exponentials? I have a `do_integral` flag which can be used to suppress symbolical integration. I use the `Integral(..)` function for that. Is there anything like that for matrix exponentials? The Symbolic function can be very slow (it uses diagonalization?) or even unable to compute the exponential, raising an error.

benepla committed:
Various fixes in StateSpaceModel.evaluate() method.

- Removed trailing dps keyword in numerical evalutation method
- Added 'simplify' keyword argument / changed keywird argument handling
- fixed 'do_integrals' behaviour (if False, it now returns a Matrix of
  Integral objects and not an Integral object with matrix valued
  integrand)
- fixed a typo

benepla committed:
Removed scipy usage from evaluate method of StateSpaceModel.

Numercial evaluation now uses mpmath.quad() which is slower, but now
travis should pass.You can still use numpy arrays as input.

I fixed two minor thing, I fogot during the API change. the getAttr()
handler now works again, as well as the cascade() and parralel()
methods.

benepla committed:
Added default behaviour if no or less than four matrcies are given.

StateSpaceModel() returns a Model
StaeSpaceModel[[0, 0],
			   [0, 0]]
StateSpaceModel(A) returns
StateSpaceModel[[A, 0],
				[0, 0]]
and so on..
The Error behaviour stayed the same.

benepla committed:
Added parallel() and cascade() methods to TransferFunctionModel

benepla committed:
Added a __eq__() method for StateSpaceModel and TranferFunctionModel

Both representations can be compared to each other so that two systems
are equal if they produce the same output, when given the same input and
initial conditions. For Example:

In [1]: ssm1 = StateSpaceModel(
        Matrix([[0, 1],
               [-a0, -a1]]),
        Matrix([0, 1]),
        Matrix([[b0, b1]]),
        zeros(1))

In [2]: tfm = TransferFunctionModel(ssm1)

In [3]: ssm2 = StateSpaceModel(tfm)

In [4]: ssm1 == tfm
Out[4]: True

In [5]: ssm2 == tfm
Out[5]: True

In [6]: ssm1 == ssm2
Out[6]: True

although the systems ssm1 and ssm2 look different.

benepla committed:
Added unittests for all methods except numerical evaluation

I added a unittest folder which cntains tests for all functionalities
of the module except numerical evaluation. The tests include:

test_StateSpaceModel_create()
test_TransferFunctionModel_create()
test_StateSpaceModel_eval()
test_controllability_matrix()
test_StateSpaceModel_cascade()
test_StateSpaceModel_parallel()
test_TransferFunctionModel_parallel()
test_equality()

All tests pass on my machine.

m3zz0m1x commented:
I added a unittests folder, just like in other modules. Do I have to do anything, so that they will be executed alongside with the other tests of sympy (so write the test file in some global list or so)?

benepla committed:
Added a __init__.py file in the controlsys/test directory

benepla committed:
corrected unittests.

An old Version was comited in the former commit, so that not all tests
passed. This Commit fixed that issue.

benepla committed:
Changed string representation of both models and added __repr__()

Before the commit, there was no representation in a normal python
console:

>>> A, B, C, D = Matrix([1,2]), Matrix([2,3]), Matrix([2]), Matrix([0])
>>> StateSpaceModel(A, B, C, D)
<sympy.controlsys.lti.StateSpaceModel object at 0x00000000043C7630>

The new representation looks like this:

>>> A, B, C, D = Matrix([1,2]), Matrix([2,3]), Matrix([2]), Matrix([0])
>>> StateSpaceModel(A, B, C, D)
StateSpaceModel(
Matrix([
[1],
[2]]),
Matrix([
[2],
[3]]),
Matrix([[2]]),
Matrix([[0]]))

moorepants commented:
The tests run based on the naming convention of the directories, files, functions, and classes. Basically, things should start with "test" or "Test".  We use the py.test tool suite for testing. Looks like you've done thing correctly. You can run the tests from the sympy root with `bin/test`. Do `bin/test -h` for options, for things like only running tests in one module, etc.

benepla committed:
Added doctest for all methods of StateSpace- and TransferFunctionMoel

m3zz0m1x commented:
Okay the tests work and pass

benepla committed:
Fixed python3 compatibility and removed print in evaluate method

the module used xrange() at some points. I changed these to the range
funciton from sympy.core.compatibility

There was a problem with lcm(map( .. )) in python 3. The lcm() function
(least common multiple) can work with a list but not a generator object,
which map() returns in python 3.
I fixed this issue by using lcm(list(map( .. )))

I removed the printing of timesteps in the evaluate method.

The tests now all pass on my machine in python 2.7 and python 3.4

benepla committed:
Added two tests for numerical evaluation.

m3zz0m1x commented:
I added docstring examples and two tests for numerical evaluation. All tests pass on my machine in python 3.4 and 2.7 on my machine and so does Travis. 
From my point of view, there is nothing left to do here =)
Still I think that this is a lot of code, and someone else will have to a look at this in detail before merging. Or do you think different @moorepants ?

jksuom commented on sympy/controlsys/__init__.py:
```diff
@@ -0,0 +1,22 @@
+"""controlsys is a package for control systems theory. It currently only
+supports linear, time invariant systems via the module 'lti'. You can
+create linear control systems in state space or transfer funciton model
+representation, transform theese representatin into oneanother,
```
Typos on 2 lines.

jksuom commented on sympy/controlsys/lti.py:
```diff
@@ -0,0 +1,1171 @@
+            raise TypeError("'representation' must have at least 4 matrix-valued entries")
+
+    def _find_realization(self, G, s):
+        """ Represenatation [A, B, C, D] of the state space model
```
Representation

jksuom commented on sympy/controlsys/lti.py:
```diff
@@ -0,0 +1,1171 @@
+        return sol
+
+    #
+    # _solve_numericaly
```
numerically?

jksuom commented on sympy/controlsys/lti.py:
```diff
@@ -0,0 +1,1171 @@
+    def cascade(self, anotherSystem):
+        """ Returns the cascade interconnection of the system and another system
+
+        The casade interconnection of two systems P1 and P2 is the system for which
```
cascade

jksuom commented on sympy/controlsys/lti.py:
```diff
@@ -0,0 +1,1171 @@
+    ==========
+
+    arg : StateSpaceModel, Matrix
+        the state space model to contruct the transfer function model from, or the transfer matrix G
```
construct

jksuom commented on sympy/controlsys/lti.py:
```diff
@@ -0,0 +1,1171 @@
+class TransferFunctionModel(object):
+    """ Transfer function model of a linear, time invariant crontrol system
+
+    Represents the transfere Function model with a transfer function Matrix G in laplace space.
```
transfer function model with a transfer function matrix G

laplace -> Laplace

jksuom commented on sympy/controlsys/lti.py:
```diff
@@ -0,0 +1,1171 @@
+        # return result
+        return self.G.subs(self.s, s) * u
+
+    def cascade(self, anotherSystem):
```
Perhaps anotherSystem -> other ?

benepla committed:
fixed various typos, renamed second argument of interconnection methods

The second argument was named 'anotherSystem' before and is now simply
called 'other', as in __eq__(self, other).

Various typos were fixed in docstrings, comments etc.

m3zz0m1x commented:
Fixed lots of typos (I just noticed that you can press F6 in sublime to enable spell checking :stuck_out_tongue_winking_eye: ) and renamed the argument of the interconnection methods cascade and parallel.

jksuom commented on sympy/controlsys/lti.py:
```diff
@@ -0,0 +1,1171 @@
+    """
+    if strict is False:
+        return all(degree(en.as_numer_denom()[0], s) <=
+                   degree(en.as_numer_denom()[1], s) for en in m)
```
It looks like ``as_numer_denom`` would be evaluated twice for each entry. That could be avoided by defining the degree of an entry along the following lines and then writing ``deg(en) <= 0`` or ``deg(en) < 0``.
```
    def deg(en):
        nd = en.as_numer_denom()
        return degree(nd[0]) - degree(nd[1])
```

benepla committed:
Reduced calls of as_numer_denom in _is_proper.

The method was called twice for each entry. I fixed this, using the
proposal of jksuom.
An additional function _entry_deg() was introduced which first splits
the entry into numerator and denominator and then calculates the
difference of their degrees.

benepla committed:
Edited the former commit slightly, so that all tests pass.

One problem was that another together is needed now for all tests to
pass. I added this.
Another problem was, that I forgot to pass a Symbol to the _entry_deg()
function. This caused an error when entrys were constants. This is
fixed now.

benepla committed:
fixed string representation of TransferFunctionModel so that doctests pass

The __str__() method now uses sstr() to get the Matrix instead of str()

certik commented:
Can you add the tutorial from this PR description as a module docstring?

czgdp1807 commented:
It looks like a good PR and can be extended.
@certik Is there anything else to be done other than addressing [this comment](https://github.com/sympy/sympy/pull/9916#issuecomment-165241817)? 
ping @jksuom Do you have any comments for adding/modifying something for the new PR.

certik commented:
@czgdp1807 I think that's it. I don't have any other objections.

czgdp1807 commented:
Closing this in favour of #17866
